source/core/slang-name-convention-util.cpp

```diff
@@ -0,0 +1,193 @@
+
+#include "slang-name-convention-util.h"
+
+#include "slang-char-util.h"
+#include "slang-string-util.h"
+
+namespace Slang
+{
+
+/* static */void NameConventionUtil::split(NameConvention convention, const UnownedStringSlice& slice, List<UnownedStringSlice>& out)
+{
+    switch (convention)
+    {
+        case NameConvention::Kabab:
+        {
+            StringUtil::split(slice, '-', out);
+            break;
+        }
+        case NameConvention::Snake:
+        {
+            StringUtil::split(slice, '_', out);
+            break;
+        }
+        case NameConvention::Camel:
+        {
+            typedef CharUtil::Flags CharFlags;
+            typedef CharUtil::Flag CharFlag;
+
+            CharFlags prevFlags = 0;
+            const char*const end = slice.end();
+
+            const char* start = slice.begin();
+            for (const char* cur = start; cur < end; ++cur)
+            {
+                char c = *cur;
+                const CharUtil::Flags flags = CharUtil::getFlags(c);
+
+                if (flags & CharFlag::Upper)
+                {
+                    if (prevFlags & CharFlag::Lower)
+                    {
+                        // If we go from lower to upper, we have a transition
+                        out.add(UnownedStringSlice(start, cur - 1));
+                        start = cur;
+                    }
+                    else if ((prevFlags & CharFlag::Upper) && cur + 1 < end)
+                    {
+                        // TODO(JS): This doesn't catch situations where there is a single letter like
+                        // ABox, because of this rule will be split to AB, ox
+                        // If we have acronyms not being capitalized this would not be an issue
+
+                        // Could be an acronym, if the next character is lower
+                        if (CharUtil::isLower(cur[1]))
+                        {
+                            out.add(UnownedStringSlice(start, cur));
+                            start = cur;
+                        }
```

Comment:
I'm not sure I follow this comment, because the obvious splitting is `A`, `Box`.

The informal rules seem to be:

* If the previous char was lower-case, and the next is upper-case, split. (you definitely have that rule)

* If the previous char char was upper-case, and the next is upper-case, and the one after *that* is lowercase, split. (That seems to be what you implemented here)

In the case of `ABox`, we'd have the point where the `A` has been read, so the cursor is placed like `A|Box`. In that case, `flags & CharFlag::Upper` is true, `prevFlags & CharFlag::Upper` is true, `cur + 1 < end` is true, and `CharUtil::isLower(cur[1])` should be true as well. It seems like that would split at the `|` (`cur`) and lead to us getting `A`, `Box` as the output, which is what we'd want.

Am I misunderstanding the code, or the comment?

Reply:
The comments not right - the limitation around upper acronyms is two acronyms next to one another. Like say IAABBSystem.